drivers/regulator/regulator_fixed.c

```diff
@@ -0,0 +1,392 @@
+	}
+
+	rc = gpio_pin_set(data->gpio, cfg->gpio_pin, false);
+
+#ifdef CONFIG_MULTITHREADING
+	if (rc == -EWOULDBLOCK) {
+		/* Perform the disable and finalization in a work
+		 * item.
+		 */
+		LOG_DBG("%s: stop deferred", cfg->regulator_name);
+		__ASSERT_NO_MSG(data->task == WORK_TASK_UNDEFINED);
+		data->task = WORK_TASK_DISABLE;
+		data->notify = notify;
+		k_work_submit(&data->delayed_work.work);
+		return;
+	}
+#endif /* CONFIG_MULTITHREADING */
+
+finalize:
+	finalize_transition(data, notify, delay_us, rc);
+
+	return;
+}
+
+static int enable_onoff(const struct device *dev, struct onoff_client *cli)
+{
+	struct driver_data_onoff *data = dev->data;
+
+	return onoff_request(&data->mgr, cli);
+}
+
+static int disable_onoff(const struct device *dev)
+{
+	struct driver_data_onoff *data = dev->data;
+
+	return onoff_release(&data->mgr);
+}
+
+static const struct onoff_transitions transitions =
+	ONOFF_TRANSITIONS_INITIALIZER(start, stop, NULL);
+
+static const struct regulator_driver_api api_onoff = {
+	.enable = enable_onoff,
+	.disable = disable_onoff,
+};
+
+static int regulator_fixed_init_onoff(const struct device *dev)
+{
+	struct driver_data_onoff *data = dev->data;
+	int rc = common_init(dev, &data->gpio);
+
+	data->dev = dev;
+	onoff_manager_init(&data->mgr, &transitions);
+#ifdef CONFIG_MULTITHREADING
+	k_delayed_work_init(&data->delayed_work, onoff_worker);
+#endif /* CONFIG_MULTITHREADING */
+
+	if (rc >= 0) {
+		rc = 0;
+	}
+
+	LOG_INF("%s onoff: %d", dev->name, rc);
+
+	return rc;
+}
+
+struct driver_data_sync {
+	const struct device *gpio;
+	struct onoff_sync_service srv;
+};
+
+#if DT_HAS_COMPAT_STATUS_OKAY(regulator_fixed_sync) - 0
+
+static int enable_sync(const struct device *dev, struct onoff_client *cli)
+{
+	struct driver_data_sync *data = dev->data;
+	const struct driver_config *cfg = dev->config;
+	k_spinlock_key_t key;
+	int rc = onoff_sync_lock(&data->srv, &key);
+
+	if ((rc == 0)
+	    && ((cfg->options & OPTION_ALWAYS_ON) == 0)) {
+		rc = gpio_pin_set(data->gpio, cfg->gpio_pin, true);
+	}
+
+	return onoff_sync_finalize(&data->srv, key, cli, rc, true);
+}
+
+static int disable_sync(const struct device *dev)
+{
+	struct driver_data_sync *data = dev->data;
+	const struct driver_config *cfg = dev->config;
+	k_spinlock_key_t key;
+	int rc = onoff_sync_lock(&data->srv, &key);
+
+	if  ((cfg->options & OPTION_ALWAYS_ON) != 0) {
+		rc = 0;
+	} else if (rc == 1) {
+		rc = gpio_pin_set(data->gpio, cfg->gpio_pin, false);
+	} else if (rc == 0) {
+		rc = -EINVAL;
+	} /* else rc > 0, leave it on */
+
+	return onoff_sync_finalize(&data->srv, key, NULL, rc, false);
+}
+
+static const struct regulator_driver_api api_sync = {
+	.enable = enable_sync,
+	.disable = disable_sync,
+};
+
+static int regulator_fixed_init_sync(const struct device *dev)
+{
+	struct driver_data_sync *data = dev->data;
+	const struct driver_config *cfg = dev->config;
+	int rc = common_init(dev, &data->gpio);
+
+	(void)regulator_fixed_init_onoff;
```

Comment:
What are these three lines for?

Reply:
They eliminated "function defined but not referenced" in cases where only `_sync` regulators are present.

I may be able to remove them with preprocessor conditional that prevent them from being defined, but I think I tried that and it got ugly.

Reply:
How about __maybe_unused ?

Reply:
Possibly; no precedent in zephyr.